### PR TITLE
Always emit leading 'var' declarations for synthesized namespaces

### LINKED ISSUE #17596: Synthesized ModuleDeclarations are missing the top level variable declaration in the emit
**TypeScript Version:**  2.4.2

**Context**
Needed by Angular / Tsickle for converting TypeScript types into closure type comments.

Our workaround (see [transformer_util.ts](https://github.com/angular/tsickle/blob/master/src/transformer_util.ts)):
- copy the `ts.Node.Symbol` from the original node over to the synthetic node (note that this is an `@internal` field)

**Code**

Plunker is [here](https://plnkr.co/edit/Xg9s2gADI7AoEXLJ56CK?p=preview).

case.ts
```ts
module Reflect {
  const x = 1;
}
```

transformer.ts:
```
function forceSyntheticModuleDeclaration(context: ts.TransformationContext) {
  return (sourceFile: ts.SourceFile): ts.SourceFile => {
    return visitNode(sourceFile);

    function visitNode<T extends ts.Node>(node: T) {
      if (node.kind === ts.SyntaxKind.ModuleBlock) {
        const md = node as ts.ModuleBlock;
        return ts.updateModuleBlock(md, ts.setTextRange(ts.createNodeArray([...md.statements]));
      }
      return ts.visitEachChild(node, visitNode, context); 
    }
  };
}
```

**Expected behavior:**
A top level variable statement should be emitted:
```js
var Reflect;
(function (Reflect) {
    var x = 1;
})(Reflect || (Reflect = {}));
```

**Actual behavior:**
No top level variable statement is emitted:
```js
(function (Reflect) {
    var x = 1;
})(Reflect || (Reflect = {}));
```

/cc @mprobst @evmar @alexeagle

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Down-leveling a namespace to ES involved an optimization within individual files to avoid emitting multiple `var` declarations by grabbing the namespace's symbol's name and checking whether a prior symbol with a given name already had a `var` declaration emitted. However, the check was not conservative enough to account for synthetic nodes that don't have symbols.

~~With this PR, we unconditionally emit the `var` declaration for namespaces that have no symbols.~~

With this PR, we just use local identifiers to keep track of declarations in the current scope.

Fixes #17596.

### CODE REVIEW

File: src/compiler/transformers/ts.ts [hunk 1]
```diff
@@ -2657,18 +2654,22 @@ namespace ts {
...
-         * Determines whether a declaration is the first declaration with the same name emitted
-         * in the current scope.
+         * Determines whether a declaration is *could* be the first declaration with
+         * the same name emitted in the current scope. Only returns false if we are absolutely
+         * certain a previous declaration has been emitted.
          */
-        function isFirstEmittedDeclarationInScope(node: Node) {
+        function isPotentiallyFirstEmittedDeclarationInScope(node: Node) {
+            // If the node has a named symbol, then we have enough knowledge to determine
+            // whether a prior declaration has been emitted.
```
Comment (rbuckton) [inline src/compiler/transformers/ts.ts]: We should consider using `getParseTreeNode` here and get the symbol of the parse tree node. If there is no parse tree node, or the node does not have a symbol, we should fall back to the text of the declaration's name.

**Comment by DanielRosenwasser** [comment]
I've changed the implementation to just never use symbols; we have `escapedName` anyway.

File: src/compiler/transformers/ts.ts [hunk 2]
```diff
@@ -2639,36 +2649,40 @@ namespace ts {
...
+            const name = declaredNameInScope(node);
+            if (!currentScopeFirstDeclarationsOfName.has(name)) {
+                currentScopeFirstDeclarationsOfName.set(name, node);
             }
         }
 
         /**
-         * Determines whether a declaration is the first declaration with the same name emitted
-         * in the current scope.
+         * Determines whether a declaration is *could* be the first declaration with
```
Comment (rbuckton) [inline src/compiler/transformers/ts.ts]: > a declaration ~is~ \*could\*

File: src/compiler/transformers/ts.ts [hunk 3]
```diff
@@ -2639,36 +2649,40 @@ namespace ts {
...
+                return currentScopeFirstDeclarationsOfName.get(name) === node;
             }
 
-            return false;
+            // Otherwise, we can't be sure. For example, this node could be synthetic.
+            return true;
+        }
+
+        function declaredNameInScope(node: FunctionDeclaration | ClassDeclaration | ModuleDeclaration | EnumDeclaration): __String {
+            if (node.name.kind !== SyntaxKind.Identifier) {
```
Comment (rbuckton) [inline src/compiler/transformers/ts.ts]: `Debug.assertNode(node.name, isIdentifier)`

File: src/compiler/transformers/ts.ts [hunk 4]
```diff
@@ -2746,7 +2760,7 @@ namespace ts {
                 return createNotEmittedStatement(node);
             }
 
-            Debug.assert(isIdentifier(node.name), "TypeScript module should have an Identifier name.");
+            Debug.assert(isIdentifier(node.name), "A TypeScript namespace should have an Identifier name.");
```
Comment (rbuckton) [inline src/compiler/transformers/ts.ts]: `Debug.assertNode(node, isIdentifier)`

**Pull request merged**